File: algorithms/dynamic-programming/sos-dp/typescript/sosDp.ts
export function sosDp(n: number, f: number[]): number[] {
    const size = 1 << n;
    const sos = [...f];

    for (let i = 0; i < n; i++) {
        for (let mask = 0; mask < size; mask++) {
            if (mask & (1 << i)) {
                sos[mask] += sos[mask ^ (1 << i)];
            }
        }
    }
    return sos;
}

const readline = require('readline');
const rl = readline.createInterface({ input: process.stdin });
const lines: string[] = [];
rl.on('line', (line: string) => lines.push(line.trim()));
rl.on('close', () => {
    const n = parseInt(lines[0]);
    const f = lines[1].split(' ').map(Number);
    const result = sosDp(n, f);
    console.log(result.join(' '));
});

File: algorithms/dynamic-programming/sos-dp/python/sos_dp.py
import sys

def sos_dp(n, f):
    """Compute sum over subsets for each bitmask."""
    sos = f[:]
    for i in range(n):
        for mask in range(1 << n):
            if mask & (1 << i):
                sos[mask] += sos[mask ^ (1 << i)]
    return sos


if __name__ == "__main__":
    data = sys.stdin.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    f = [int(data[idx + i]) for i in range(1 << n)]
    result = sos_dp(n, f)
    print(' '.join(map(str, result)))
